### PR TITLE
feat: make int arithmetic overflow a runtime error

### LINKED ISSUE #544: enhancement: `int` arithmetic overflow wraps silently without error
## Summary

Arithmetic on `int` (i64) silently wraps on overflow, which can cause hard-to-debug issues. Low-level types have `checked_add` etc., but the default `int` type has no protection.

## Reproduction

```ry
big = 9_223_372_036_854_775_807   # INT64_MAX
print(big + 1)                     # -9223372036854775808 (silent wrap)
```

## Context

- Low-level types (`i32`, `u8`, etc.) have `checked_add`, `saturating_add`, `wrapping_add` — explicit overflow control
- The default `int` type (i64) wraps silently, which is inconsistent with Ry's safety-oriented design

## Suggestion

Consider one of:
1. **Runtime error** on overflow for `int` (like Swift's default behavior)
2. **Compile-time warning** when overflow is possible
3. **Document the behavior** explicitly if silent wrapping is intentional

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

- `int` 型の算術演算 (`+`, `-`, `*`, 単項 `-`) でオーバーフローが発生した場合、サイレントにラップする代わりにランタイムエラーを発生させるようにした
- 定数式のオーバーフローはコンパイル時に検出される
- 低レベル型 (`i32`, `u8` 等) は影響を受けず、従来通りラップ動作を維持

Closes #544

## Test plan

- [x] C++ death tests: add/sub/mul/neg overflow → exit(1) with "integer overflow"
- [x] C++ tests: constant folding overflow → compile-time error
- [x] C++ tests: normal arithmetic still works (no false positives)
- [x] C++ tests: compound assignment (`+=`) overflow detection
- [x] C++ tests: low-level types still wrap
- [x] Ry spec tests: normal int arithmetic (7 cases)
- [x] All existing tests pass (1310 C++ tests, 74 spec files)
- [x] ASan clean (no memory issues)

## Summary by CodeRabbit

* **New Features**
  * High-level integer arithmetic (`+`, `-`, `*`, unary `-`) now raises a runtime error on overflow; constant-expression overflows are detected at compile time.

* **Documentation**
  * Updated type and operator guides to clarify new overflow semantics and document explicit helpers for checked, saturating, and wrapping behavior.

* **Tests**
  * Added runtime and compile-time tests covering overflow, boundary cases, compound assignments, and preserved low-level wraparound behavior.

### CODE REVIEW

File: docs/reference/types.md [hunk 1]
```diff
@@ -717,4 +717,5 @@ result = add_one(v)   # any(int) is unwrapped to int; result is 43
...
+- **`int` arithmetic overflow is a runtime error** -- Arithmetic (`+`, `-`, `*`, unary `-`) on the high-level `int` type raises a runtime error on overflow, similar to Swift's default behavior. This prevents silent data corruption from two's complement wrapping. Constant expressions that overflow are caught at compile time.
+- **Low-level integer overflow wraps around** -- Arithmetic on low-level integer types uses two's complement wrapping on overflow (signed) or modular arithmetic (unsigned). For example, `i32` max value `2147483647 + 1` wraps to `-2147483648`. This matches C behavior. For explicit overflow control, use `checked_add/sub/mul` (returns `Result<T, Error>`), `saturating_add/sub/mul` (clamps to type bounds), or `wrapping_add/sub/mul` (self-documenting wrapping). See [Function Reference](functions.md#checkedsaturating-arithmetic).
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply docs/reference/types.md]: Fixed: Removed the misleading "This matches C behavior" phrasing. Now describes it as "Ry-defined two's complement wrapping" since signed overflow is undefined behavior in C.

[the remark this answers is not in the document]
Comment (t0k0sh1) [reply docs/reference/types.md]: Fixed: Changed example from `2147483647 + 1` to `2147483647i32 + 1i32` to make it explicitly low-level.

File: src/codegen_expr.cpp [hunk 2]
```diff
@@ -640,9 +649,20 @@ llvm::Value *CodeGen::emitArithmeticOp(const std::string &op, llvm::Value *lhs,
...
+        if (op == "*") return emitIntOverflowCheck(llvm::Intrinsic::smul_with_overflow, lhs, rhs, "mul");
+    }
+    // Low-level i64/u64 fallthrough: wrap and propagate metadata
+    auto propagateHint = [&](llvm::Value *result) -> llvm::Value* {
+        if (!llNameHint.empty()) low_level_type_names_[result] = llNameHint;
+        return result;
+    };
+    if (op == "+") return propagateHint(builder_.CreateAdd(lhs, rhs, "add"));
+    if (op == "-") return propagateHint(builder_.CreateSub(lhs, rhs, "sub"));
+    if (op == "*") return propagateHint(builder_.CreateMul(lhs, rhs, "mul"));
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp]: Acknowledged. The root cause is the pre-existing ConstantInt metadata limitation (#311) — `checkLowLevelTypeMix` cannot distinguish `int` from `i64` for constant expressions. This is tracked as a separate bug in #595, which will address per-operand low-level intent detection. Not fixing in this PR to avoid scope creep.

File: src/codegen_arith.cpp [hunk 3]
```diff
@@ -113,6 +113,48 @@ llvm::Value *CodeGen::emitSaturatingArithmetic(const std::string &callee,
...
+            llvm::APInt a = cl->getValue();
+            llvm::APInt b = cr->getValue();
+            bool overflow = false;
+            llvm::APInt result;
+            if (intrinsicId == llvm::Intrinsic::sadd_with_overflow)
+                result = a.sadd_ov(b, overflow);
+            else if (intrinsicId == llvm::Intrinsic::ssub_with_overflow)
+                result = a.ssub_ov(b, overflow);
+            else
+                result = a.smul_ov(b, overflow);
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_arith.cpp]: Fixed: Added explicit `smul_with_overflow` check and `codegenError` for unexpected intrinsic IDs.

File: src/codegen_expr.cpp [hunk 4]
```diff
@@ -152,6 +152,15 @@ llvm::Value *CodeGen::emitExprVariant(const std::unique_ptr<UnaryExpr> &e) {
...
+        // Overflow check only for high-level int; low-level i64 wraps
+        bool isLowLevel = isLowLevelTy(val);
+        if (!isLowLevel) {
+            if (auto *ne = std::get_if<NumberExpr>(&e->operand->data))
+                isLowLevel = isLowLevelTypeName(ne->suffix);
+        }
+        if (val->getType() == i64Ty_ && !isLowLevel)
+            return emitIntOverflowCheck(llvm::Intrinsic::ssub_with_overflow,
+                                         llvm::ConstantInt::get(i64Ty_, 0), val, "neg");
         return builder_.CreateNeg(val, "neg");
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp]: This is the same ConstantInt metadata limitation (#311) already raised in the previous review round by Copilot. Adding a Constant guard here breaks i64 wrapping behavior (verified by test failure). The proper fix — making `getExprLowLevelSuffix` recursive and preserving per-operand low-level intent — is tracked in #595.
